Framework/DataHandling/inc/MantidDataHandling/PulseIndexer.h

```diff
@@ -0,0 +1,61 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2024 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#pragma once
+
+#include "MantidDataHandling/DllConfig.h"
+
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace Mantid {
+namespace DataHandling {
+
+/** PulseIndexer : TODO: DESCRIPTION
+ */
+class MANTID_DATAHANDLING_DLL PulseIndexer {
+public:
+  PulseIndexer(std::shared_ptr<std::vector<uint64_t>> event_index, const std::size_t firstEventIndex,
+               const std::size_t numEvents, const std::string &entry_name);
+
+  /// Which element in the event_index array is the one to use
+  size_t getFirstPulseIndex(const size_t event_index) const;
+  /// The range of event(tof,detid) indices to read for this pulse
+  std::pair<size_t, size_t> getEventIndexRange(const size_t pulseIndex) const;
+  /**
+   * The first event(tof,detid) index to read for this pulse.
+   * This is only public for testing.
+   */
+  size_t getStartEventIndex(const size_t pulseIndex) const;
+  /**
+   * The one past the last event(tof,detid) index to read for this pulse
+   * This is only public for testing.
+   */
+  size_t getStopEventIndex(const size_t pulseIndex) const;
+
+private:
+  PulseIndexer(); // do not allow empty constructor
+
+  /// vector of event index (length of # of pulses)
+  const std::shared_ptr<std::vector<uint64_t>> m_event_index;
```

Comment:
am I to understand here that `m_event_index[3]=42` means that event with index `42` is the first event recorded for pulse with index `3`? If so, I'd rather read a more descriptive docstring